weblab/fitting/views.py

```diff
@@ -197,3 +204,150 @@ def get(self, request, *args, **kwargs):
         }
 
         return JsonResponse(response)
+
+
+class FittingResultCreateView(LoginRequiredMixin, PermissionRequiredMixin, UserFormKwargsMixin, FormView):
+    """
+    Create and submit a fitting result from models, protocols, fitting specs and datasets
+    and (where relevant) their versions.
+    """
+    permission_required = 'fitting.run_fits'
+    form_class = FittingResultCreateForm
+
+    template_name = 'fitting/fittingresult_create_form.html'
+
+    def get_initial(self):
+        initial = super().get_initial()
+        if 'model' in self.request.GET:
+            initial['model'] = get_object_or_404(ModelEntity, pk=self.request.GET['model'])
+        elif 'protocol' in self.request.GET:
+            initial['protocol'] = get_object_or_404(ProtocolEntity, pk=self.request.GET['protocol'])
+        elif 'fittingspec' in self.request.GET:
+            initial['fittingspec'] = get_object_or_404(FittingSpec, pk=self.request.GET['fittingspec'])
+        elif 'dataset' in self.request.GET:
+            initial['dataset'] = get_object_or_404(Dataset, pk=self.request.GET['dataset'])
+
+        return initial
+
+    def post(self, request, *args, **kwargs):
+        form = self.get_form()
+        if form.is_valid():
+            return self.form_valid(form)
+        else:
+            return self.form_invalid(form)
+
+    def form_valid(self, form):
+        self.runnable, is_new = submit_fitting(
+            form.cleaned_data['model_version'],
+            form.cleaned_data['protocol_version'],
+            form.cleaned_data['fittingspec_version'],
+            form.cleaned_data['dataset'],
+            self.request.user,
+            True,
+        )
+
+        queued = self.runnable.status == FittingResultVersion.STATUS_QUEUED
+
+        if is_new:
+            if queued:
+                messages.info(self.request, "Fitting experiment submitted to the queue.")
+            else:
+                messages.error(self.request, "Fitting experiment could not be run: " + self.runnable.return_text)
+        else:
+            messages.info(self.request, "Fitting experiment was already run.")
+
+        return super().form_valid(form)
+
+    def get_success_url(self):
+        if hasattr(self, 'runnable'):
```

Comment:
Is it possible for `self.runnable` not to exist? And what happens if so?

Reply:
I'm not sure why I added that check - if it gets to this call then `self.runnable` will have been assigned in `form_valid`. I'll remove it.